File: dotnet/Tests/Steps/ExampleSteps.cs
using Microsoft.Playwright;

namespace QAItest.Playwright.DotNet.Tests.Steps;

public static class ExampleSteps
{
    public sealed class ExampleContext
    {
        public IPage Page { get; init; } = default!;
        public bool Executed { get; set; }
    }

    public static ExampleContext Initialize(IPage page) => new() { Page = page, Executed = false };

    public static async Task OpenHomepageAsync(ExampleContext context)
    {
        context.Executed = true;
        await context.Page.GotoAsync("https://playwright.dev/");
    }

    public static async Task ExpectHomepageAsync(ExampleContext context)
    {
        await context.Page.GetByRole(AriaRole.Link, new() { Name = "Get started" }).WaitForAsync();
        if (!context.Executed)
        {
            throw new InvalidOperationException("Scenario did not execute.");
        }
    }
}

File: java/src/test/java/com/qaitest/playwright/steps/ExampleSteps.java
package com.qaitest.playwright.steps;

import com.microsoft.playwright.Page;

public final class ExampleSteps {
  private ExampleSteps() {}

  public static ExampleContext initialize(Page page) {
    return new ExampleContext(page, false);
  }

  public static void openHomepage(ExampleContext context) {
    context.executed = true;
    context.page.navigate("https://playwright.dev/");
  }

  public static void expectHomepage(ExampleContext context) {
    context.page.getByRole(com.microsoft.playwright.options.AriaRole.LINK,
      new Page.GetByRoleOptions().setName("Get started")).waitFor();
    if (!context.executed) {
      throw new IllegalStateException("Scenario did not execute");
    }
  }

  public static final class ExampleContext {
    public final Page page;
    public boolean executed;

    public ExampleContext(Page page, boolean executed) {
      this.page = page;
      this.executed = executed;
    }
  }
}
